### DISCUSSION TITLE
Proper way to close a connection

### DISCUSSION THREAD

**Discussion by brianreinhold** [pyusb/pyusb#432] [Q&A] [answered]
I thought I posted this question last Friday but I must have done something wrong.

I am wondering how to properly close a connection. There is no 'open()' or 'close()' method in the API as far as I can tell. It seems like the connection is 'opened' by ```Device.set_configuration()``` method. Currently I close the connection at the end of the program by calling
```usb.util.dispose_resources(Device)```.

The problem is that when I plug in my device, run my application and leave the device plugged in, and run a second time, the device does not behave properly (it still communicates and transfers data) as if it is left in a bad state. If I run the application a third time, it runs okay. It turns out that if I leave the device plugged in, the application runs correctly every other time. 

On the other hand, if I unplug the device and plug it in before each run, the application run correctly every time.

If I switch to the usb1 library (which has explicit Open() and Close() methods) the device runs correctly every time even if it remains plugged in. So I feel I must not be shutting down the device correctly upon program termination.

Now the problem may be due to some other error I am making. The two libraries are different enough that the two applications are not a simple switch. Thus I ask about proper shutdown in PyUSB. If what I am doing now for shutdown in PyUSB is correct, then I need to look elsewhere for the problem.

Can anyone tell me what the proper shutdown procedure is upon application termination, OR if there is a problem re-running a PyUSB application with the device remaining plugged in.

Note that I tried calling Device.reset() before releasing resources and I get an error 
```
Traceback (most recent call last):
  File "/home/pi/Public/usb_connect.py", line 331, in <module>
    main()
  File "/home/pi/Public/usb_connect.py", line 321, in main
    device.reset()
  File "/home/pi/Public/usb/core.py", line 961, in reset
    self._ctx.backend.reset_device(self._ctx.handle)
  File "/home/pi/Public/usb/backend/libusb1.py", line 911, in reset_device
    _check(self.lib.libusb_reset_device(dev_handle.handle))
  File "/home/pi/Public/usb/backend/libusb1.py", line 604, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 2] Entity not found
```
Though the last thing I do before calling this method is read the current time on the device and that works. **On the plus calling this method DOES solve my problem - the application runs correctly every time while remaining plugged in between runs**, but that exception bothers me. I can clean up the code by catching the exception but that only hides the issue.

**Comment by jonasmalacofilho** [comment, marked as answer]
> It seems like the connection is 'opened' by `Device.set_configuration()` method.

Actually, the closest to "opening the connection" is claiming the interface, which you can do explicitly or PyUSB will do for you when needed.

On the other hand, `set_configuration()`, as the name implies, is about configuring the device.  Technically, you don't need to call it just to access an interface... provided the device has already been configured (in the way you expect), possibly by the OS.

(So, in practice, you really might need to care about `set_confguration()`... but it's tricky, as I mention bellow).

> Currently I close the connection at the end of the program by calling `usb.util.dispose_resources(Device)`.

That's right.

---

While I can't really tell what's wrong from the information you provided so far, I have a suspicion it might be related to `set_configuration()`: are you aware of the issue of calling it on already configured devices?

If not, take a look at our [corresponding FAQ entry](https://github.com/pyusb/pyusb/blob/master/docs/faq.rst#how-not-to-call-set_configuration-on-a-device-already-configured-with-the-selected-configuration), as well as the LibUSB documentation referenced there.

**Reply by brianreinhold** [reply]
Jonas, I did not know about that case and I added the check into my code. The config was already set. However, it did not change the result. I still get the exception at the end when calling reset() but everything works.

I tried calling device.get_active_configuration().close() instead of reset() at the close but that method threw an exception saying that there was no attribute 'close' even though intellisense showed me the method. So I keep the active config check in the code and keep the reset and deal with the 'Entity not found' exception.

My shutdown (dispose resources still functions well)
```
    try:
        print("Closing connection")
        device.reset()
    except usb.core.USBError as e:
        print(e)
    usb.util.dispose_resources(device)
    led_opts.reset()
    ep_out = None
    ep_in = None
    device = None
    dev_info = None
```
Here is my method that gets the bulk read and write endpoints. It's riddled with print statements as I cannot line by line debug.

```
def get_bulk_endpoints(devs:list[usb.core.Device]) -> tuple[usb.core.Endpoint, usb.core.Endpoint, usb.core.Device]:
    global ep_in, ep_out, device
    # Parses the descriptor information obtained from pyUSB and prints the results
    # But then it finds the bulk in and out interfaces and sets them. The method
    # returns a tuple containing the two endpoints and the Device
    for dev in devs:
        #print(dev)
        print("    Number of Configurations: ", dev.bNumConfigurations)
        print("    Serial number String Descriptor index: ", dev.iSerialNumber)
        print("    Number of configurations: ", len(dev.configurations()))
        cfgs = dev.configurations()
        for cfg in cfgs:
            intfs: tuple[usb.core.Interface, ...] = cfg.interfaces()
            print("    Number of interfaces: ", len(intfs))
            for intf in intfs:
                print("    Interface number: ", intf.bInterfaceNumber)
                endpts:list[usb.core.Endpoint] = intf.endpoints()
                print("    Number of endpoints: ", len(endpts))
                for endpt in endpts:
                    if usb.util.endpoint_type(endpt.bmAttributes) == usb.util.ENDPOINT_TYPE_BULK:
                        if usb.util.endpoint_direction(endpt.bEndpointAddress) == usb.util.ENDPOINT_IN:
                            ep_in = endpt
                            print("    Bulk endpoint  in found")
                        else:
                            ep_out = endpt
                            print("    Bulk endpoint out found")
                if ep_in and ep_out:
                    try:
                        print("Get current active config")
                        curr_cfg = dev.get_active_configuration()
                    except usb.core.USBError:
                        curr_cfg = None
                        print("Interface not claimed yet")
                    if curr_cfg is None or curr_cfg.bConfigurationValue != cfg.bConfigurationValue:
                        print("Claiming interface")
                        dev.set_configuration(cfg.bConfigurationValue)
                    else:
                        print ("Active config already set")
                    break;
            if ep_in and ep_out:
                device = dev
                break

        print("    Product: ", dev.product)
        print("    Manufacturer: ", dev.manufacturer)
        print("    Serial number: ", dev.serial_number)
    return ep_in, ep_out, device
```
and the log output of running that method:
```
Number of devices  1
    Number of Configurations:  1
    Serial number String Descriptor index:  3
    Number of configurations:  1
    Number of interfaces:  1
    Interface number:  0
    Number of endpoints:  2
    Bulk endpoint out found
    Bulk endpoint  in found
Get current active config
Active config already set
    Product:  Study Watch
    Manufacturer:  Verily
    Serial number:  C2QT510X2052004F
Both bulk in and bulk out endpoints found
```

**Reply by jonasmalacofilho** [reply]
> However, it did not change the result. I still get the exception at the end when calling reset() but everything works.

But do you still _need_ the `reset()` after that fix?

Resetting the device is prone to causing errors, as during & after it it might not be possible for all expectations to be held in sync across all layers of the stack.  By the way, here's the LibUSB documentation for [`libusb_reset_device()`](https://libusb.sourceforge.io/api-1.0/group__libusb__dev.html#gafee9c4638f1713ca5faa867948878111):

> Perform a USB port reset to reinitialize a device. 
> 
> The system will attempt to restore the previous configuration and alternate settings after the reset has completed.
> 
> **If the reset fails, the descriptors change, or the previous state cannot be restored, the device will appear to be disconnected and reconnected. This means that the device handle is no longer valid (you should close it) and rediscover the device. A return code of LIBUSB_ERROR_NOT_FOUND indicates when this is the case.**

(`LIBUSB_ERROR_NOT_FOUND` is the corresponding error code for the "Entity not found" error)[^1]

[^1]: https://libusb.sourceforge.io/api-1.0/group__libusb__misc.html#ggab2323aa0f04bc22038e7e1740b2f29efa28214194dc4399e51777408b2c27f03f

---

> Here is my method that gets the bulk read and write endpoints. It's riddled with print statements as I cannot line by line debug.

I was only able to glance at it, sorry.

But if you haven't heard of it yet, Python has an interactive debugger: [pdb](https://docs.python.org/3/library/pdb.html).  You can have it start at some convenient point by adding a call to the `breakpoint()` built-in function.

Also, the messages

```py
print("Interface not claimed yet")
print("Claiming interface")
```

are not accurate, as there you're checking the configuration, which has to do with the _whole_ device, not a particular interface (and your code wouldn't be claiming it anyway).

**Reply by brianreinhold** [reply]
I still find that I have to call the reset method even after the check in order for the device to behave correctly when reconnected. That is why I was looking for the proper way to close an existing connection. In the original libusb1 package there was a close method that was used. I don't know what it did under the hood but the device reconnected without issue in both cases (leaving the device plugged in and unplugging the device between restarts).

The print statements were just quick text to tell me which case got invoked. I added them yesterday morning as I put in your piece of code to check the configuration state. Barring the print statements, I think I am making the correct calls to create a connection and use the bulk reads and writes. Everything seems to work okay except for that exception and it would be nice if I could find a proper shutdown sequence such that the device reconnects nice.

Note that if I unplug the device between restarts of the app, I don't need the reset for it to behave correctly on a reconnect. The reset is only needed to get a functional restart when I leave the watch connected.

**Reply by jonasmalacofilho** [reply]
The most obvious difference I can think of is that `libusb_close()` will not be called as long as you keep a reference to the `Device` instance.  Maybe that's somehow related to the

> the device does not behave properly (it still communicates and transfers data) as if it is left in a bad state

issue you're seeing?  But that would probably suggest some other problem in your implementation of the device protocol, that somehow gets accidentally corrected by LibUSB and/or the kernel in some but all situations.

(I've seen some behavior somewhat resembling this before: in one instance, the device required some padding, but that was automatically fixed by the backend and kernel in some platforms, but not others... and when the padding was missing, the device would still work for a while, but erratically... it's also the type of thing that could be affected by a backend discarding vs. reusing buffers).
